src/stdune/bin.ml

```diff
@@ -24,24 +24,21 @@ let exists fn =
   | exception (Unix.Unix_error _) -> false
   | _ -> true
 
-let best_prog dir prog =
-  let fn = Path.relative dir (prog ^ ".opt" ^ exe) in
-  if exists fn then
-    Some fn
+let add_exe prog =
+  if String.is_suffix prog ~suffix:exe then
+    prog
   else
-    let fn = Path.relative dir (prog ^ exe) in
-    if exists fn then
-      Some fn
-    else
-      None
+    prog ^ exe
 
 let which ~path prog =
   let rec search = function
     | [] -> None
     | dir :: rest ->
-      match best_prog dir prog with
-      | None -> search rest
-      | Some fn -> Some fn
+      let fn = Path.relative dir (add_exe prog) in
```

Comment:
We can do the `add_exe` just once outside the loop

Reply:
Indeed. Fixed.